src/Compilers/CSharp/Portable/Compiler/MethodCompiler.cs

```diff
@@ -930,7 +930,7 @@ public override object VisitField(FieldSymbol symbol, TypeCompilationState argum
                     var lazySemanticModel = body == null ? null : new Lazy<SemanticModel>(() =>
                     {
                         var syntax = body.Syntax;
-                        var semanticModel = (CSharpSemanticModel)_compilation.GetSemanticModel(syntax.SyntaxTree);
+                        var semanticModel = (CSharpSemanticModel)AnalyzerDriver.GetOrCreateCachedSemanticModel(syntax.SyntaxTree, _compilation, _cancellationToken);
```

Comment:
@mavasani If it belongs anywhere, I believe the change belongs in `AnalyzerDriver` - which should provide an alternate mechanism for populating the weak `SemanticModel` of a `CompilationEvent` when it has been cleared by the collector.

Not that any of this will help you. The compiler will still have to recreate the nodes for the method body by rebinding it. Reusing an existing semantic model doesn't avoid that, it just moves the work from one place to another.